synapse/storage/databases/main/client_ips.py

```diff
@@ -538,15 +599,20 @@ async def get_last_client_ip_by_device(
         """
         ret = await super().get_last_client_ip_by_device(user_id, device_id)
 
-        # Update what is retrieved from the database with data which is pending insertion.
+        # Update what is retrieved from the database with data which is pending
+        # insertion, as if it has already been stored in the database.
         for key in self._batch_row_update:
-            uid, access_token, ip = key
+            uid, _access_token, ip = key
             if uid == user_id:
                 user_agent, did, last_seen = self._batch_row_update[key]
+
+                if did is None:
+                    # These updates don't make it to the `devices` table
+                    continue
+
                 if not device_id or did == device_id:
-                    ret[(user_id, device_id)] = {
+                    ret[(user_id, did)] = {
```

Comment:
Is this a bugfix? (If `device_id` is falsey, should we be using it as part of the key?)

Reply:
It is a bugfix, from #10970  

Basically the data store buffers updates in `_batch_row_update` and writes them to the database periodically. The call to `super().get_last_client_ip_by_device(...)` above fetches from database while this code adds data from the in-memory buffer to the results.

The db implementation of `get_last_client_ip_by_device` always uses the correct `device_id` in the key whereas the original code here didn't.